sdk/src/main/scala/io/horizen/certificatesubmitter/AbstractCertificateSubmitter.scala

```diff
@@ -408,13 +409,30 @@ abstract class AbstractCertificateSubmitter[
                   proofWithQuality.getValue
                 } try to send it to mainchain")
 
-                mainchainChannel.sendCertificate(certificateRequest) match {
-                  case Success(certificate) =>
-                    log.info(s"Backward transfer certificate response had been received. Cert hash = " + BytesUtils.toHexString(certificate.certificateId))
-
-                  case Failure(ex) =>
-                    log.error("Creation of backward transfer certificate had been failed.", ex)
+                // check if there are already certificates present in mempool for current or higher epoch
+                var isCertificateAlreadyPresent = false
+                mainchainChannel.getTopQualityCertificates(BytesUtils.toHexString(BytesUtils.reverseBytes(params.sidechainId))) match {
+                  case Success(mcRefTry) =>
+                    (mcRefTry.mempoolCertInfo, mcRefTry.chainCertInfo) match {
```

Comment:
Here we are supposing to have both field (mempoolCertInfo and chainCertInfo) populated, are we sure that this is always the case?

Reply:
Match cases expanded to include those when only mempoolCertInfo or chainCertInfo is populated
Upon closer look, I'm not sure if it's always the case. Looking at [this ](https://github.com/HorizenOfficial/Sidechains-SDK/blob/2196d647be114599c1c035c41cb3b3b03b69d3a5/sdk/src/main/scala/io/horizen/certificatesubmitter/strategies/CeasingSidechain.scala#L64)call of getTopQaulityChainCertificates, there are cases when only one of them is taken into consideration

Reply:
Match clauses also throw an exception if none of the branches match. Hence the compiler warning if a match clause is "non-exhaustive". We should always cover all cases, even if it is a "catch all" with `_ => ..` to throw an exception. This will remove the warning and make the error explicit. Improves readability a lot, imho.